### DISCUSSION TITLE
Playwright post-demo discussion outline

### DISCUSSION THREAD

**Discussion by DukeFerdinand** [AlaskaAirlines/auro-library#96] [Announcements]
This outline is intended to act as a semi-comprehensive guide to using playwright from scratch in an auro component.

### **Step 1. External tooling**
To avoid duplicating a lot of small, tedious copy-paste work across dozens of repositories, either the new Auro CLI tool or auro-library repositories need to be updated to be made aware of playwright configuration files.

As auro-library is already configured to synchronize files, a quick solution might be to add another processor flow to sync a remote `playwright.config.ts` file to the local repository.

### **Step 2. Preparing a repository** 
At a minimum, a component repository needs the playwright config file, the "@playwright/test" npm dependency, and one test to run. Normally, this would be accomplished by running `npm init playwright@latest` but this is not strictly required. The `playwright@latest` initializer is just a wizard to walk through the creation of a new playwright config file with configured defaults.

Here's a link to the main playwright installation docs: https://playwright.dev/docs/intro

Since the configuration files will be stored externally and referenced per-repo, the minimum sync process (or creation process in the case of the spec.ts file - this should not be overwritten) should ensure the following:
- package.json includes `"@playwright/test": "^1.49.0",`
- some github action configured to run playwright tests ([see here](https://playwright.dev/docs/ci-intro))
- `e2e-tests/<component-name>.spec.ts` exists and contains the following default snapshot test
```typescript
import { test, expect } from '@playwright/test';  
  
test.afterEach(async ({ page }, testInfo) => {  
  if (testInfo.status !== testInfo.expectedStatus) {  
    // Get a unique place for the screenshot.  
    const screenshotPath = testInfo.outputPath(`failure.png`);  
    // Add it to the report.  
    testInfo.attachments.push({ name: 'screenshot', path: screenshotPath, contentType: 'image/png' });  
    // Take the screenshot itself.  
    await page.screenshot({ path: screenshotPath, timeout: 5000 });  
  }  
});  
  
test('regression test', async ({ page }) => {  
    await page.goto('/demo');  
  
    const screenshot = await page.screenshot({  
        fullPage: true,  
    });  
  
    // expect screenshot to match snapshot  
    expect(screenshot).toMatchSnapshot('button.png');  
});
```
---

- `playwright.config.ts` exists and includes the following
```typescript
import { defineConfig, devices } from '@playwright/test';  
  
/**  
 * Read environment variables from file. * https://github.com/motdotla/dotenv */// import dotenv from 'dotenv';  
// import path from 'path';  
// dotenv.config({ path: path.resolve(__dirname, '.env') });  
  
/**  
 * See https://playwright.dev/docs/test-configuration. */export default defineConfig({  
  testDir: './e2e-tests',  
  /* Run tests in files in parallel */  
  fullyParallel: true,  
  /* Fail the build on CI if you accidentally left test.only in the source code. */  
  forbidOnly: !!process.env.CI,  
  /* Retry on CI only */  
  retries: process.env.CI ? 2 : 0,  
  /* Opt out of parallel tests on CI. */  
  workers: process.env.CI ? 1 : undefined,  
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */  
  reporter: 'html',  
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */  
  use: {  
    /* Base URL to use in actions like `await page.goto('/')`. */  
    baseURL: 'http://127.0.0.1:8000/',  
  
    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */  
    trace: 'on-first-retry',  
  },  
  
  /* Configure projects for major browsers */  
  projects: [  
    {  
      name: 'chromium',  
      use: { ...devices['Desktop Chrome'] },  
    },  
  
    {  
      name: 'firefox',  
      use: { ...devices['Desktop Firefox'] },  
    },  
  
    {  
      name: 'webkit',  
      use: { ...devices['Desktop Safari'] },  
    },  
  
    /* Test against mobile viewports. */  
    // {    //   name: 'Mobile Chrome',    //   use: { ...devices['Pixel 5'] },    // },    // {    //   name: 'Mobile Safari',    //   use: { ...devices['iPhone 12'] },    // },  
    /* Test against branded browsers. */    // {    //   name: 'Microsoft Edge',    //   use: { ...devices['Desktop Edge'], channel: 'msedge' },    // },    // {    //   name: 'Google Chrome',    //   use: { ...devices['Desktop Chrome'], channel: 'chrome' },    // },  ],  
  
  /* Run your local dev server before starting the tests */  
  webServer: {  
    command: 'npm run serve',  
    url: 'http://127.0.0.1:8000/demo',  
    reuseExistingServer: !process.env.CI,  
  },  
});
```

### Step 3. Writing a new test
Tests can be written totally manually (and probably will) for most simple test cases. For more complex workflows like using a date picker, clicking multiple things, or triggering various features with keybinds, playwright has a "record" feature for generating tests based on events that are fired during the recording phase.

*Doug's note: I'm normally a "manual testing" fan, but this looks like an awesome tool!*

[See here for more](https://playwright.dev/docs/codegen)

### Step 4. CI Integration
See the link above mentioning CI integration. Each repository needs a github action file to run these tests, and further will require some experimentation to check what sort of developer experience is surfaced when a test fails.

### Required changes
Unfortunately, due to the way Auro example HTML is consumed in existing workflows - being injected into one large entrypoint HTML file, there is not an easy way to test specific examples for regression. In short, each HTML file needs to be injected into a more complete "parent shell" that registers the target web component(s), imports CSS, and performs any required setup.

Why? Playwright uses a real browser behind the scenes, and real browsers can't fake a DOM like React testing library, Vite, and similar test suites.

*Doug's note: I would hope that this is an additive action and doesn't require fundamentally changing how the docs are built. It's just HTML templates, and injecting HTML templates has been solved hundreds of ways over the years*

### Closing thoughts
I'm probably missing something (or many things) in the implementation process, but overall the Playwright team has done an AWESOME job at making this tool easy to use. When I first used this tool years ago, it was nowhere near this easy to use!

**Comment by chrisfalaska** [comment]
What were some of the other tools or frameworks you mentioned previously that integrate with Playwright? 

Is that to say, extending the general Playwright functionality is a possibility in the future?

Would incorporating a tool like Axe for accessibility testing be considered a launch feature, or is it something planned for a fast follow-up?: https://www.npmjs.com/package/axe-playwright / https://playwright.dev/docs/accessibility-testing
